### PR TITLE
tests: migrate assertions to be more modern

### PR DESCRIPTION
**Pull request by cyphar**
Quite a lot of tests (especially in `integration-cli`) use boolean assertions which don't provide any useful information about why a comparison failed. This is already kind of annoying when debugging a failing test, but when running the integration suite on an actual machine (where the source code parsing doesn't work) this makes it almost impossible to figure out what the actual error is.

A lot of this is due to limitations of `gotest.tools` (see https://github.com/gotestyourself/gotest.tools/issues/147 for instance) but we can work around most of these issues.

**NOTE**: I've only tackled the "easy" cases for now. All of the transformations were done using `gofmt -r` or `golang.org/x/tools/cmd/eg`, which should make it easier to verify that the tests weren't subtly broken by these changes.

Future work:
- [ ] Fix all of the assertions that have help messages (these are less critical because usually the message gives you some kind of information). Unfortunately, `gofmt` cannot handle these automatically, and we would need to run `gofmt` for every number of arguments used. We could use `sed` but this could lead to bugs, and `eg` seems to be able to do this but it is incredibly poorly documented.
- [x] Migrate `assert.Assert(t, errors.Is(...))` to `assert.ErrorIs` or `is.ErrorIs`.
- [x] Migrate `assert.Assert(t, is.Contains(err.Error(), ...)` to `assert.ErrorContains`.
- [ ] Fix inverse comparisons. `integration-cli/checker` doesn't help with this entirely (see https://github.com/gotestyourself/gotest.tools/issues/147) -- we will instead need to create a new `cmpext` package that will let us create a generic inverter that has useful error messages. This will then let us tackle these issues:
  - [ ] `assert.Assert(t, a != nil)`
  - [ ] `assert.Assert(t, !strings.Contains(...))`
  - [ ] ...

### CODE REVIEW

**Comment by cyphar** [comment]
I'll work on the `cmpext` stuff if there is any interest in taking this kind of patchset.

File: integration-cli/docker_cli_prune_unix_test.go
Comment (thaJeztah) [inline integration-cli/docker_cli_prune_unix_test.go:108]: Not necessarily for this PR, unless you need to make more changes; looks like these `strings.Trimspace()` calls here (and couple of other places below?) are redundant, as we're checking if it _contains_ `id`

File: integration-cli/docker_cli_run_test.go
Comment (thaJeztah) [inline integration-cli/docker_cli_run_test.go:2285]: Note to self: these blanket "any error should do" cases keep catching my eye. I may have a go at finding these to see if we can be more specific (admitted, the output check below should probably prevent catching random unrelated errors, but still)

File: integration-cli/docker_cli_build_test.go
Comment (thaJeztah) [inline integration-cli/docker_cli_build_test.go:4637]: Still looking for more elegant solutions to these (negative matches); we could probably consider putting `result.Stdout()` as part of a custom failure message to help (definitely not needed for this PR though)


edit: ah; I see you mentioned this in your PR description as well

> Fix inverse comparisons. integration-cli/checker doesn't help with this entirely

**Comment by thaJeztah** [comment]
Looks like potentially flaky test; I see this one was mentioned in https://github.com/moby/moby/pull/48609 as being flaky (and possibly to be fixed with that one?) cc @robmry FYI

```
=== Failed
=== FAIL: amd64.integration.networking TestAccessPublishedPortFromHost/userland-proxy=true/IPv6=false (1.09s)
    port_mapping_linux_test.go:392: assertion failed: error is not nil: Get "http://127.0.0.1:1234": EOF
    --- FAIL: TestAccessPublishedPortFromHost/userland-proxy=true/IPv6=false (1.09s)

```

**Review by thaJeztah — APPROVED** [review]
Thanks for working on this! I think I had some branches stashed with some of these changes, but never got to complete them 🤗 


changes so far LGTM (assuming CI passes)

I see you left some TODO's, notably;

> Fix all of the assertions that have help messages


I'd be fine merging this one as-is, and work on remaining ones as follow-up(s); perhaps some of those changes (if we can't come with a "smart" solution to find/replace) would be just a matter of some elbow-grease to go through them? (should be something we'd only have to do once to bring it up to speed).

File: integration-cli/docker_cli_run_test.go
Comment (cyphar) [reply integration-cli/docker_cli_run_test.go:2285]: Yeah these checks really should be `ErrorsIs` or at the very least `ErrorContains` but unfortunately that migration can't be easily automated. Luckily there's fewer than 100 of them, so we could migrate them in one patch:

```
% git grep "assert.\(Assert\|Check\)(.*err.* != nil" | wc -l
87
```

File: integration-cli/docker_cli_build_test.go
Comment (cyphar) [reply integration-cli/docker_cli_build_test.go:4637]: Yeah, the trick is that we need to create a custom type that wraps `assert.BoolOrComparison` which then inverts `Success` as well as providing a wrapped `FormatMessage` that describes that the comparison was inverted. I suspect the wrapped description is going to be quite janky (it would probably look like the suggestion that the author of `gotest.tools` proposed in https://github.com/gotestyourself/gotest.tools/issues/147#issuecomment-622200454). Unfortunately, `FormatMessage` is technically internal to `gotest.tools` so it could break in the future (but the author suggested doing it this way, so I'm not sure).

Since it seems you'd be happy with a patch like that, I can try to cook one up.

Comment (thaJeztah) [reply integration-cli/docker_cli_build_test.go:4637]: Yeah, a custom helper could work still, but perhaps we could still look if we could contribute a patch to `gotest.tools` that would be acceptable to their maintainers 🤔 


For the formatting, we could still go the manual way;

```go
assert.Assert(c, !strings.Contains(result.Stdout(), "foo"), "Expected: %q to contain %q", result.Stdout(), "foo")
```

Which (if all of these are in the same (`integration-cli`) package could be a local helper I guess;

```go
func assertNotContains(t testing.T, val, expected string) {
    t.Helper()
    assert.Assert(t, !strings.Contains(val, expected), "Expected %q not to contain %q", val, expected)
}
```

**Comment by thaJeztah** [comment]
@cyphar let me know if you want to add more changes to this PR before merging, or if you're good with bring the current set of changes in already. I think the current set of changes are already an improvement, so (from my perspective) the remaining ones are not a blocker for bringing those in already.

File: integration-cli/docker_cli_build_test.go
Comment (cyphar) [reply integration-cli/docker_cli_build_test.go:4637]: That would be the simplest option if we want to avoid over-complicating things. I'll go with that version now and we can do the more complicated version in the future...

**Comment by cyphar** [comment]
I'd like to fix the `errors.Is` and `strings.Contains(err.Error(), ...)` checks but all of the other stuff can be done in later PRs IMHO.

**Comment by cyphar** [comment]
A proper `ErrorsIs` might actually be more involved because a lot of the cases that look ripe for conversion might not actually wrap the error properly or the test might be checking something subtly different. So I've just gone with the "dumb" version of the patch where we only convert cases that are clearly correct (so I've skipped converting any `errors.As` code as well as more complicated checks).

**Comment by cyphar** [comment]
@thaJeztah I think this should be okay for now.

**Review by thaJeztah — APPROVED** [review]
LGTM, thank you!

**Review by neersighted — APPROVED** [review]

**Pull request merged**
